modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStream.java

```diff
@@ -9,145 +9,155 @@
+    private static final StorageRetryStrategy STORAGE_RETRY_STRATEGY = GoogleCloudStorageService.createStorageRetryStrategy();
 
-    static final int MAX_SUPPRESSED_EXCEPTIONS = 10;
-
-    private final OperationPurpose purpose;
-    private final MeteredStorage client;
-    private final BlobId blobId;
-    private final long start;
-    private final long end;
-    private final int maxAttempts;
-    private InputStream currentStream;
-    private int attempt = 1;
-    private List<StorageException> failures = new ArrayList<>(MAX_SUPPRESSED_EXCEPTIONS);
-    private long currentOffset;
-    private boolean closed;
-    private Long lastGeneration;
+    GoogleCloudStorageRetryingInputStream(GoogleCloudStorageBlobStore blobStore, OperationPurpose purpose, BlobId blobId)
+        throws IOException {
+        this(blobStore, purpose, blobId, 0, Long.MAX_VALUE - 1);
+    }
 
-    // Used for testing only
-    GoogleCloudStorageRetryingInputStream(OperationPurpose purpose, MeteredStorage client, BlobId blobId) throws IOException {
-        this(purpose, client, blobId, 0, Long.MAX_VALUE - 1);
+    GoogleCloudStorageRetryingInputStream(
+        GoogleCloudStorageBlobStore blobStore,
+        OperationPurpose purpose,
+        BlobId blobId,
+        long start,
+        long end
+    ) throws IOException {
+        super(new GoogleCloudStorageBlobStoreServices(blobStore, purpose, blobId), purpose, start, end);
     }
 
-    // Used for testing only
-    GoogleCloudStorageRetryingInputStream(OperationPurpose purpose, MeteredStorage client, BlobId blobId, long start, long end)
-        throws IOException {
-        if (start < 0L) {
-            throw new IllegalArgumentException("start must be non-negative");
-        }
-        if (end < start || end == Long.MAX_VALUE) {
-            throw new IllegalArgumentException("end must be >= start and not Long.MAX_VALUE");
+    private static class GoogleCloudStorageBlobStoreServices implements BlobStoreServices<Long> {
+
+        private final GoogleCloudStorageBlobStore blobStore;
+        private final OperationPurpose purpose;
+        private final BlobId blobId;
+
+        private GoogleCloudStorageBlobStoreServices(GoogleCloudStorageBlobStore blobStore, OperationPurpose purpose, BlobId blobId) {
+            this.blobStore = blobStore;
+            this.purpose = purpose;
+            this.blobId = blobId;
         }
-        this.purpose = purpose;
-        this.client = client;
-        this.blobId = blobId;
-        this.start = start;
-        this.end = end;
-        this.maxAttempts = client.getOptions().getRetrySettings().getMaxAttempts();
-        this.currentStream = openStream();
-    }
 
-    private InputStream openStream() throws IOException {
-        try {
+        @Override
+        public SingleAttemptInputStream<Long> getInputStream(@Nullable Long lastGeneration, long start, long end) throws IOException {
+            final MeteredStorage client = blobStore.clientNoRetries();
```

Comment:
I am not sure whether this is necessary for now. The current S3 implementation also has nested retries, inner layer with provided by the SDK and outer layer with our own implementation. That seems to be the existing behaviour of GCS retrying input stream as well. Maybe we should keep it as is for now. I am slightly concerned that this reduces number of total retries in production.

We can change it in future for all CSP together if it is necessary.

Reply:
I think there's a slight difference between the CSPs here. S3 doesn't support resuming a streaming download when it fails part-way through, at least [not the way we invoke it](https://github.com/aws/aws-sdk-java/issues/856#issuecomment-2015713550). So while we have wrapped retries for failures that occur on the initial request, we only have a single "resume-on-fail" layer in S3, implemented in the `RetryingInputStream`.

GCS does support resuming on failure part-way through an input stream read, so if we wrap the client behaviour we'll be adding a layer of retries here. Perhaps that's the least-bad option? I agree it would be bad to reduce the amount of retries in either case as part of this change. I'll see what we can do.